src/zjs_fs.c

```diff
@@ -0,0 +1,940 @@
+        ERR_PRINT("first parameter must be path string\n");
+        return ZJS_UNDEFINED;
+    }
+    if (!jerry_value_is_string(argv[1])) {
+        ERR_PRINT("second parameter must be flags string\n");
+        return ZJS_UNDEFINED;
+    }
+    if (async) {
+        if (!jerry_value_is_function(argv[argc - 1])) {
+            ERR_PRINT("last parameter must be callback function\n");
+            return ZJS_UNDEFINED;
+        }
+    }
+
+    file_handle* handle = new_file();
+    handle->fd_val = jerry_create_object();
+
+    jerry_size_t jlen = jerry_get_string_size(argv[0]);
+    char path[jlen + 1];
+
+    int wlen = jerry_string_to_char_buffer(argv[0], (jerry_char_t *)path, jlen);
+    if (jlen != wlen) {
+        ERR_PRINT("size mismatch\n");
+        return ZJS_UNDEFINED;
+    }
+    path[wlen] = '\0';
+
+    jlen = jerry_get_string_size(argv[1]);
+    char mode[jlen + 1];
+
+    wlen = jerry_string_to_char_buffer(argv[1], (jerry_char_t *)mode, jlen);
+    if (jlen != wlen) {
+        ERR_PRINT("size mismatch\n");
+        return ZJS_UNDEFINED;
+    }
+    mode[wlen] = '\0';
+
+    DBG_PRINT("Opening file: %s, mode: %s\n", path, mode);
+
+    handle->mode = get_mode(mode);
+
+    /*
+     * TODO: Currently Zephyr has no concept of file permissions, when you open
+     *       a file it is automatically readable and writable. This if block
+     *       is here in case things change and we can throw an exception if a
+     *       file is opened with improper permissions.
+     */
+    if (file_exists(path)) {
+        if (handle->mode & FLAGS_NO_CREATE) {
+            DBG_PRINT("Opening file %s in mode %s throws exception if file exists\n", path, mode);
+            // TODO: Throw exception
+        } else {
+            DBG_PRINT("Opening existing file %s with mode %s\n", path, mode);
+        }
+    } else {
+        if (handle->mode & FLAGS_CREATE) {
+            DBG_PRINT("Opening file %s in mode %s throws exception if file does not exists\n", path, mode);
+            // TODO: Throw exception
+        } else {
+            DBG_PRINT("Creating new file %s\n", path);
+        }
+    }
+
+    handle->error = fs_open(&handle->fp, path);
+    if (handle->error != 0) {
+        DBG_PRINT("could not open file: %s\n", path);
+    }
+
+    jerry_set_object_native_handle(handle->fd_val, (uintptr_t)handle, NULL);
+
+    if (async) {
+        zjs_callback_id id = zjs_add_callback_once(argv[argc - 1],
+                                                   this,
+                                                   handle,
+                                                   NULL);
+
+        jerry_value_t args[2];
+
+        args[0] = jerry_create_number(handle->error);
+        args[1] = handle->fd_val;
+
+        zjs_signal_callback(id, args, sizeof(jerry_value_t) * 2);
+
+        return ZJS_UNDEFINED;
+    }
+
+    return handle->fd_val;
+}
+
+static jerry_value_t zjs_open_sync(const jerry_value_t function_obj,
```

Comment:
We discussed this but node "default" is async. The latest PR defaults to use xxxSync() as all the API's to match Node, and all the async API's are compiled out.